File: src/main/java/io/github/ajoz/workshop/fp/java/part_4/solutions/exercise_6/Exercise6.java
package io.github.ajoz.workshop.fp.java.part_4.solutions.exercise_6;

import io.github.ajoz.workshop.fp.java.tools.Function1;
import io.github.ajoz.workshop.fp.java.tools.Function2;
import io.github.ajoz.workshop.fp.java.tools.Try;

import java.util.LinkedList;
import java.util.List;
import java.util.NoSuchElementException;

import static io.github.ajoz.workshop.fp.java.part_4.solutions.exercise_6.Flow.cycle;
import static io.github.ajoz.workshop.fp.java.part_4.solutions.exercise_6.Flow.generate;
import static java.util.Arrays.asList;

interface Flow<A> {
    Try<A> next();

    default <B, C> Flow<C> zip(final Flow<B> other,
                               final Function2<A, B, C> zipper) {
        return new ZippingFlow<>(this, other, zipper);
    }

    default Flow<A> take(final int threshold) {
        return new TakeFlow<>(this, threshold);
    }

    default List<A> toList() {
        final List<A> list = new LinkedList<>();
        while (true) {
            final Try<A> element = next();
            if (element.isFailure())
                break;

            list.add(element.get());
        }
        return list;
    }

    static <A> Flow<A> cycle(final List<A> list) {
        return new CycleListFlow<>(list);
    }

    static <A> Flow<A> generate(final A seed,
                                final Function1<A, A> generator) {
        return new Seed1Flow<>(seed, generator);
    }
}

class ZippingFlow<A, B, C> implements Flow<C> {
    private final Flow<A> left;
    private final Flow<B> right;
    private final Function2<A, B, C> zipper;

    ZippingFlow(final Flow<A> left,
                final Flow<B> right,
                final Function2<A, B, C> zipper) {
        this.left = left;
        this.right = right;
        this.zipper = zipper;
    }

    @Override
    public Try<C> next() {
        final Try<A> nextLeft = left.next();
        if (nextLeft.isFailure())
            return Try.failure(new NoSuchElementException("Left seq is out of elements to zip!"));

        final Try<B> nextRight = right.next();
        if (nextRight.isFailure())
            return Try.failure(new NoSuchElementException("Right seq is out of elements to zip!"));

        return nextLeft.flatMap(a -> nextRight.flatMap(b -> Try.success(zipper.apply(a, b))));
    }
}

class CycleListFlow<A> implements Flow<A> {
    private final List<A> list;
    private int current;

    CycleListFlow(final List<A> list) {
        this.list = list;
    }

    @Override
    public Try<A> next() {
        if (list.size() == 0)
            return Try.failure(new NoSuchElementException("No elements in this Flow!"));

        if (current >= list.size())
            current = 0;

        final Try<A> next = Try.success(list.get(current));
        current++;
        return next;
    }
}

class Seed1Flow<A> implements Flow<A> {
    private final Function1<A, A> generator;
    private A seed;

    public Seed1Flow(final A seed,
                     final Function1<A, A> generator) {
        this.seed = seed;
        this.generator = generator;
    }

    @Override
    public Try<A> next() {
        final Try<A> next = Try.success(seed);
        seed = generator.apply(seed);
        return next;
    }
}

class TakeFlow<A> implements Flow<A> {
    private final Flow<A> upstream;
    private final int amount;

    private int taken;

    TakeFlow(final Flow<A> upstream,
             final int amount) {
        this.upstream = upstream;
        this.amount = amount;
    }

    @Override
    public Try<A> next() {
        // take an element from the upstream Flow
        final Try<A> next = upstream.next();
        // if the elemenent does not exist then just propagate the failure
        if (next.isFailure())
            return next;

        // if already taken enough elements then just propagate the failure
        if (taken >= amount) {
            return Try.failure(new NoSuchElementException("Reached the Flow amount: " + amount));
        }

        // increment the currently taken amount
        taken++;
        // return the upstream Flow item
        return next;
    }
}

public class Exercise6 {
    public static void main(final String[] args) {
        final Flow<String> fizzFlow = cycle(asList("", "", "Fizz"));
        final Flow<String> buzzFlow = cycle(asList("", "", "", "", "Buzz"));
        final Flow<Integer> numbers = generate(1, a -> a + 1);

        final List<String> fizzBuzz =
                fizzFlow
                        .zip(buzzFlow, (a, b) -> a + b)
                        .zip(numbers, (s, integer) -> {
                            if (s.isEmpty())
                                return String.valueOf(integer);

                            return s;
                        })
                        .take(200)
                        .toList();

        System.out.println("fizzBuzz = " + fizzBuzz);
    }
}

File: kotlin/solutions/src/main/kotlin/io/github/ajoz/workshop/fp/part4/exercise6/Exercise6.kt
@file:Suppress("PackageName", "MoveLambdaOutsideParentheses")

package io.github.ajoz.workshop.fp.part4.exercise6

import io.github.ajoz.workshop.fp.part4.exercise6.Flows.cycle
import io.github.ajoz.workshop.fp.part4.exercise6.Flows.generate
import io.github.ajoz.workshop.fp.tools.control.Try
import java.util.LinkedList
import java.util.NoSuchElementException

import java.util.Arrays.asList

internal interface Flow<A> {
    fun next(): Try<A>
}

internal fun <A> Flow<A>.toList(): List<A> {
    val list = LinkedList<A>()
    while (true) {
        val element = next()
        if (element.isFailure)
            break

        list.add(element.get())
    }
    return list
}

// Part 1
internal class ZippingFlow<A, B, C>(private val left: Flow<A>,
                                    private val right: Flow<B>,
                                    private val zipper: (A, B) -> C) : Flow<C> {

    override fun next(): Try<C> {
        val nextLeft = left.next()
        if (nextLeft.isFailure)
            return Try.Failure(NoSuchElementException("Left flow is out of elements to zip!"))

        val nextRight = right.next()
        return if (nextRight.isFailure)
            Try.Failure(NoSuchElementException("Right flow is out of elements to zip!"))
        else nextLeft.flatMap { a ->
            nextRight.flatMap { b ->
                Try.Success(zipper(a, b))
            }
        }
    }
}

internal fun <A, B, C> Flow<A>.zip(other: Flow<B>,
                                   zipper: (A, B) -> C): Flow<C> =
        ZippingFlow(this, other, zipper)

// Part 2:
internal class CycleListFlow<A>(private val list: List<A>) : Flow<A> {
    private var current: Int = 0

    override fun next(): Try<A> {
        if (list.isEmpty())
            return Try.Failure(NoSuchElementException("No elements in this Flow!"))

        if (current >= list.size)
            current = 0

        val next = Try.Success(list[current])
        current++
        return next
    }
}

// Part 3:
internal class GenerateFlow<A>(private var seed: A,
                               private val generator: (A) -> A) : Flow<A> {

    override fun next(): Try<A> {
        val next = Try.Success(seed)
        seed = generator(seed)
        return next
    }
}

internal object Flows {
    fun <A> cycle(list: List<A>): Flow<A> =
            CycleListFlow(list)

    fun <A> generate(seed: A, generator: (A) -> A): Flow<A> =
            GenerateFlow(seed, generator)
}

internal class TakeFlow<A>(private val upstream: Flow<A>,
                           private val amount: Int) : Flow<A> {

    private var taken: Int = 0

    override fun next(): Try<A> {
        // take an element from the upstream Flow
        val next = upstream.next()
        // if the element does not exist then just propagate the failure
        if (next.isFailure)
            return next

        // if already taken enough elements then just propagate the failure
        if (taken >= amount) {
            return Try.Failure(NoSuchElementException("Reached the take threshold: $amount"))
        }

        // increment the currently taken amount
        taken++
        // return the upstream Flow item
        return next
    }
}

internal fun <A> Flow<A>.take(threshold: Int): Flow<A> =
        TakeFlow(this, threshold)


fun main(args: Array<String>) {
    val fizzFlow = cycle(asList("", "", "Fizz"))
    val buzzFlow = cycle(asList("", "", "", "", "Buzz"))
    val numbers = generate(1) { a -> a + 1 }

    val fizzBuzz = fizzFlow
            .zip(buzzFlow) { a, b ->
                a + b
            }
            .zip(numbers) { string, integer ->
                if (string.isEmpty())
                    "$integer"
                else
                    string
            }
            .take(200)
            .toList()

    println("fizzBuzz = $fizzBuzz")
}
